library/src/math/row.rs

```diff
@@ -158,31 +160,53 @@ impl MathRow {
 
     fn into_line_frame(self, points: &[Abs], align: Align) -> Frame {
         let ascent = self.ascent();
-        let descent = self.descent();
-        let size = Size::new(Abs::zero(), ascent + descent);
-        let mut frame = Frame::new(size);
-        let mut x = Abs::zero();
+        let mut frame = Frame::new(Size::new(Abs::zero(), ascent + self.descent()));
         frame.set_baseline(ascent);
 
-        if let (Some(&first), Align::Center) = (points.first(), align) {
-            let mut offset = first;
-            for fragment in self.iter() {
-                offset -= fragment.width();
-                if matches!(fragment, MathFragment::Align) {
-                    x = offset;
-                    break;
+        let mut next_x = {
+            let widths = if points.is_empty() || align == Align::Left {
+                Vec::new()
+            } else {
+                let mut widths = Vec::new();
+
+                let mut width = Abs::zero();
+                for fragment in self.iter() {
+                    if matches!(fragment, MathFragment::Align) {
+                        widths.push(width);
+                        width = Abs::zero();
+                    } else {
+                        width += fragment.width();
+                    }
+                }
+                widths.push(width);
+
+                widths
+            };
+
+            let mut prev_points = once(Abs::zero()).chain(points.iter().copied());
+            let mut point_widths = points.iter().copied().zip(widths);
+            let mut alternator = point_widths
+                .clone()
+                .zip(prev_points.clone())
+                .zip(LeftRightAlternator::Right);
+            move || match align {
+                Align::Left => prev_points.next(),
+                Align::Right => point_widths.next().map(|(point, width)| point - width),
+                _ => {
+                    alternator.next().map(|(((point, width), prev_point), alternator)| {
+                        match alternator {
+                            LeftRightAlternator::Left => prev_point,
+                            LeftRightAlternator::Right => point - width,
+                        }
+                    })
                 }
```

Comment:
This whole segment is a bit hard to understand. It feels to me as if this could be written in a simpler style, possibly with less iterator foo. I might be wrong though.